### ISSUE TITLE
TBB allocator integration

### ISSUE THREAD

**Issue by Im-dex** [Im-dex/xray-162#26]
Replace standard allocator by the Intel TBB scalable allocator

**Comment by jazzvaz** [comment]
@Im-dex do we really need it, and if yes, in which case? I think it is better just to use it as needed.

**Comment by Im-dex** [comment]
@jazzvaz It is useful in case when a multithreading is actively added to the engine

**Comment by jazzvaz** [comment]
@Im-dex what did you mean by replace STANDART allocator? stl type allocator or xr_memory itself?

**Comment by Im-dex** [comment]
@jazzvaz all of them, there must be only one tbb's allocator

**Comment by jazzvaz** [comment]
@Im-dex what about pools that xr_memory uses?

**Comment by Im-dex** [comment]
@jazzvaz they should all be removed

**Comment by jazzvaz** [comment]
@Im-dex performance will drop a lot

**Comment by Im-dex** [comment]
@jazzvaz why?

**Comment by jazzvaz** [comment]
@Im-dex I might be wrong, but I think gsc did good job with memory optimization. Then to remove it, tbb malloc proxy can be used instead of allocator

**Comment by Im-dex** [comment]
@jazzvaz I think Intel did a much better job than gsc :) anyway, tbb malloc already contains thread caching pools.

**Comment by jazzvaz** [comment]
@Im-dex so what is better: malloc proxy or scalable allocator? Furthermore, pool is faster with dealing many allocation and deallocation as calling malloc itself.

**Comment by Im-dex** [comment]
@jazzvaz What do you mean? scalable allocator is an allocator, malloc proxy is a way to use scalable allocator

**Comment by jazzvaz** [comment]
@Im-dex yes, that why I asked, what will be the better implementation: create an allocator or use proxy for malloc?

**Comment by Im-dex** [comment]
@jazzvaz I think proxy will be enough

**Comment by jazzvaz** [comment]
@Im-dex do we need from xr_memory: debug memory name, debug memory monitor, memory stats?

**Comment by Im-dex** [comment]
@jazzvaz nope

**Comment by jazzvaz** [comment]
@Im-dex Can I open pull request now and fill it with time as commits will be big, and you will need time to revise it?

**Comment by Im-dex** [comment]
@jazzvaz Yes, but keep in mind that it will be better to open one pr for one kind of changes (e.g. macros removing, one feature adding) cause it easier for review and makes merge faster.
